chain/network/src/peer/peer_actor.rs

```diff
@@ -1182,8 +1202,20 @@ impl PeerActor {
                     self.stop(ctx, ClosingReason::Ban(ReasonForBan::Abusive));
                 }
 
-                // Add received peers to the peer store
                 let node_id = self.network_state.config.node_id();
+
+                // Record our own IP address as observed by the peer.
+                if self.network_state.my_public_addr.read().is_none() {
+                    if let Some(my_peer_info) =
+                        direct_peers.iter().find(|peer_info| peer_info.id == node_id)
+                    {
+                        if let Some(addr) = my_peer_info.addr {
+                            let mut my_public_addr = self.network_state.my_public_addr.write();
+                            *my_public_addr = Some(addr);
```

Comment:
So if I understand correctly, we find our IP address listed in the set of peers this peer sent us and then send that in the routed state part request so that the receiver can open a tier3 connection to send the state parts? I think there's a problem here where there’s no guarantee that our public IP address looks the same to all our peers though, right?

For example, what if I run an RPC node service that has a bunch of nodes, some of which are connected to others in my fleet. Then my nodes will have some connections to nodes outside my service and some to nodes inside my service, and maybe the ones inside my service see private IP addresses for each other. And then if I hit this if branch first with one of my other nodes, I’ll be advertising a private IP address to peers that I want to send me state parts.

Also, even if this part works (which I bet it prob will in most cases), aren’t we still going to run into a problem where the node that receives this tier2 routed message and then tries to respond by initiating a tier3 message might not be able to do so if the node is behind a NAT? I think a lot of the time a NAT will only allow incoming traffic from a source that you’ve already made an outgoing connection to. If you run a node on your computer at home with your home ISP for example, then the receiving node probably won’t be able to initiate a tier3 connection to because it'll be dropped before it gets to you

Reply:
Thanks for pointing these things out.

This kind of IP discovery is also the way by which the tier2 network operates. Peers which observe the IP of the node publish a PeerInfo which is advertised to the whole network, and nodes remember only the first PeerInfo they see for a given PeerId. So if someone has the kind of setup you are describing in which they have multiple nodes in the same private network and they observe and advertise each others private IPs, they should be running into problems already. I think it doesn't happen because generally the configured boot node(s) will be outside their private network, so the first connection made will be to a node which will observe their public address correctly and advertise it to the rest of the network.

With regards to the NAT part, I think it should be OK to expect that a node wishing to participate in a blockchain protocol will accept incoming traffic from new peers. Otherwise we can imagine an extreme case in which all/nearly all nodes are behind a NAT with the restriction you are describing, and we have trouble even bootstrapping the tier2/tier1 networks due to the current way they are designed.

I would also note that the consequence of configuring a node in a way that is incompatible with this tier3 design is only that the node will not be able to obtain state parts from peers. The impact of the issue is limited to the node itself and the operator can troubleshoot the issue to get decentralized state sync working. In the worst case they don't/can't do it and will just get the parts from the external storage.

Overall I'm willing to accept the limitations of the current approach, but I'm open to iterating on these designs in the future if we can improve robustness.